Adding Cloud API

jtguggedal opened the pull request:
This PR shows the initial work on a generic Cloud API. The intention behind the API is to provide a unified way for applications to interface with different cloud solutions with having to do lots of cloud specific adjustemnts in the application code. This is now supposed to be handled by a cloud backend driver/library that provides the required cloud API functions.

The PR contains 3 commits
- Adding the API itself (note manifest pointing to branch with some additions to linker script to enable the use of `cloud_backend_get_binding()` API)
- Adding cloud API support to `nrf_cloud` lib, while still allowing the use of the lib directly.
- Switching the asset tracker to using the cloud API without dependencies `nrf_cloud`

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:577: WARNING:EMBEDDED_FUNCTION_NAME: Prefer using '"%s...", __func__' to using 'sensor_data_send', this function's name, in a string
#577: FILE: applications/asset_tracker/src/main.c:557:
+		printk("sensor_data_send failed: %d\n", err);

-:1040: WARNING:TYPO_SPELLING: 'hander' may be misspelled - perhaps 'handler'?
#1040: FILE: include/net/cloud.h:76:
+ * @brief Cloud event hander function type.

-:1056: ERROR:CODE_INDENT: code indent should use tabs where possible
#1056: FILE: include/net/cloud.h:92:
+        int (*uninit)(const struct cloud_backend *const backend);$

-:1056: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#1056: FILE: include/net/cloud.h:92:
+        int (*uninit)(const struct cloud_backend *const backend);$

-:1079: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#1079: FILE: include/net/cloud.h:115:
+ * ^I  configuration of the backend to be able to connect its online$

-:1080: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#1080: FILE: include/net/cloud.h:116:
+ * ^I  counterpart.$

-:1107: ERROR:CODE_INDENT: code indent should use tabs where possible
#1107: FILE: include/net/cloud.h:143:
+        __ASSERT_NO_MSG(backend != NULL);$

-:1107: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#1107: FILE: include/net/cloud.h:143:
+        __ASSERT_NO_MSG(backend != NULL);$

-:1109: ERROR:CODE_INDENT: code indent should use tabs where possible
#1109: FILE: include/net/cloud.h:145:
+        return backend->api->uninit(backend);$

-:1109: WARNING:LEADING_SPACE: please, no spaces at the start of a line
#1109: FILE: include/net/cloud.h:145:
+        return backend->api->uninit(backend);$

-:1114: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#1114: FILE: include/net/cloud.h:150:
+ * ^I    function returns. The socket should then be available through$

-:1115: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#1115: FILE: include/net/cloud.h:151:
+ * ^I    backend->config->socket and the application may start listening$

-:1116: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#1116: FILE: include/net/cloud.h:152:
+ * ^I    for events on it.$

-:1170: ERROR:SPACING: space required before the open parenthesis '('
#1170: FILE: include/net/cloud.h:206:
+	if(backend->api->ping != NULL) {

-:1195: WARNING:TRAILING_SEMICOLON: macros should not use a trailing semicolon
#1195: FILE: include/net/cloud.h:231:
+#define CLOUD_BACKEND_DEFINE(_name, _api)			       \
+									       \
+	static struct cloud_backend_config UTIL_CAT(_name, _config) = {		\
+		.name = STRINGIFY(_name)				       \
+	};								       \
+									       \
+	static const struct cloud_backend _name				       \
+	__attribute__ ((section(".cloud_backends"))) __attribute__((used)) =   \
+	{								       \
+		.api = &_api,						       \
+		.config = &UTIL_CAT(_name, _config)				       \
+	};

-:1197: WARNING:LONG_LINE: line over 80 characters
#1197: FILE: include/net/cloud.h:233:
+	static struct cloud_backend_config UTIL_CAT(_name, _config) = {		\

-:1197: ERROR:OPEN_BRACE: open brace '{' following function declarations go on the next line
#1197: FILE: include/net/cloud.h:233:
+	static struct cloud_backend_config UTIL_CAT(_name, _config) = {		\

-:1205: WARNING:LONG_LINE: line over 80 characters
#1205: FILE: include/net/cloud.h:241:
+		.config = &UTIL_CAT(_name, _config)				       \

-:1461: WARNING:EMBEDDED_FUNCTION_NAME: Prefer using '"%s...", __func__' to using 'uninit', this function's name, in a string
#1461: FILE: subsys/net/lib/nrf_cloud/src/nrf_cloud.c:281:
+	LOG_INF("uninit() is not implemented");

- total: 5 errors, 14 warnings, 1443 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```
## Kconfig issues

```
Undefined Kconfig symbols:

warning: undefined symbol BT_GATT_DYNAMIC_DB:

- Referenced at ../nrf/subsys/bluetooth/services/Kconfig.hids:1:

menuconfig BT_GATT_HIDS
	bool
	prompt "Human Interface Device service" if BT
	select BT_GATT_POOL if BT
	select BT_GATT_DYNAMIC_DB if BT
	depends on BT
	help
	  Enable BLE GATT Human Interface Device service.

- Referenced at ../nrf/subsys/bluetooth/services/Kconfig.lbs:1:

menuconfig BT_GATT_LBS
	bool
	prompt "Led Button service" if BT
	select BT_GATT_DYNAMIC_DB if BT
	depends on BT
	help
	  Enable Led Button service.

- Referenced at ../nrf/subsys/bluetooth/services/Kconfig.nus:1:

config BT_GATT_NUS
	bool
	prompt "Nordic UART service" if BT
	select BT_GATT_DYNAMIC_DB if BT
	depends on BT
	help
	  Enable Nordic UART service.

- Referenced at ../nrf/subsys/bluetooth/services/Kconfig.throughput:1:

config BT_GATT_THROUGHPUT
	bool
	prompt "Throughput service" if BT
	default "n" if BT
	select BT_GATT_DYNAMIC_DB if BT
	depends on BT
	help
	  Enable Nordic GATT throughput BLE service.


```

rlubos commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+	__ASSERT_NO_MSG(backend != NULL);
+
+	/* Ping will only be sent if the backend requires it. */
+	if (backend->api->ping != NULL) {
```
How about returning `-ENOTSUP` in case `ping` is not supported?

jtguggedal replied:
Yep, will fix

rlubos commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+
+#define CLOUD_BACKEND_DEFINE(_name, _api)			               \
+									       \
+	static struct cloud_backend_config UTIL_CAT(_name, _config) =	       \
```
I think we need to think thoroughly about whether this is the right direction. Technically it seems fine, but it will limit the cloud engine to a single instance. And I wonder whether it would make sense to maintain two parallel connections to the same cloud from a single device with let's say two different identities. But it might be an overcomplicated use case.

jtguggedal replied:
Yes, I'm not sure about this either. It's a hybrid of how logger and device API handle their backends and devices. Big downside is that it requires changes to linker script in the -zephyr repo, unless there's a way that I haven't found to do it another way.

The thinking behind it is that it's convenient to use some sort of `cloud_get_binding("CLOUD NAME")` call to get pointer to the backend struct, and allow mulitple backends concurrently. But as you say, only one per cloud solution. There's probably a better way to solve this.

sigvartmh replied:
Could this https://github.com/zephyrproject-rtos/zephyr/pull/15112 be used to add the relevant sections to the linker script? @jtguggedal

jtguggedal replied:
I forgot to comment here on it, but I accidently came across this solution and updated the Cloud API to use it. I placed the linker snippet within the subsys itself like this:  https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/795/files#diff-382398a27417821bcb12e888d758bda0R7 
Do you think this is an ok solution, @sigvartmh @rlubos ?

rlubos commented on applications/asset_tracker/src/main.c:
```diff
@@ -19,6 +18,12 @@
+
+#include <net/socket.h>
+
+#include <cloud_codec.h>
```
This header is local for asset tracker, should be within "", not <>.

rlubos commented on applications/asset_tracker/src/cloud_codec/cloud_codec.h:
```diff
@@ -0,0 +1,66 @@
+/*
```
Should we call it `nrf_cloud_codec`? Or is this more than nRF Cloud?

jtguggedal replied:
It basically is the `nrf_cloud_codec`, with the prefix removed to make it more generic. It wraps data in simple JSON strings so that it can be received and easily parsed by any cloud solution. It could possibly use IPSO objects or something to do it a standardized way.

metc commented on applications/asset_tracker/src/cloud_codec/cloud_codec.h:
```diff
@@ -0,0 +1,66 @@
+};
+
+int cloud_encode_sensor_data(const struct cloud_sensor_data *sensor,
+				 struct cloud_data *output);
```
Maybe adding `const` pointers for both arguments?

jtguggedal replied:
Yep, can do that

metc replied:
You could have `const pointer for both APIs.`

jtguggedal replied:
Hm, thought I added that, but seems not. Will fix it.

rlubos replied:
I don't think that making an `output` pointer `const` is a good idea. Isn't it intended to be modified?

jtguggedal replied:
I made the pointer `const`, but not the pointed data. Should work, I think

metc commented on applications/asset_tracker/src/cloud_codec/cloud_codec.c:
```diff
@@ -0,0 +1,85 @@
+	__ASSERT_NO_MSG(sensor != NULL);
+	__ASSERT_NO_MSG(sensor->data.buf != NULL);
+	__ASSERT_NO_MSG(sensor->data.len != 0);
+	__ASSERT_NO_MSG(output != NULL);
```
This is a public API, so do you really want to `assert` if a parameter is invalid?

jtguggedal replied:
I'll return error code instead

metc commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+/**
+ * @brief Cloud backend API.
+ *
+ * All cloud backends must expose this API.
```
Which of those are mandatory? I see that ping is not.

jtguggedal replied:
Initially I was thinking all except `ping`. Should be stated in the docs

metc commented on subsys/net/lib/nrf_cloud/src/nrf_cloud.c:
```diff
@@ -200,3 +201,155 @@ void nrf_cloud_process(void)
+	return 0;
+}
+
+static struct cloud_api nrf_cloud_api = {
```
This can be `const`?

metc commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+ *	  counterpart.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ * param[in] ctx Pointer to backend specific data.
```
No context parameter here, but I would say it is always good to add one, for instance as argument of the `cloud_evt_handler_t`.

jtguggedal replied:
Agree, it's the plan to add it.

metc commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+ */
+static inline int cloud_uninit(const struct cloud_backend *const backend)
+{
+	__ASSERT_NO_MSG(backend != NULL);
```
And don't you want to check if the API function is defined? I think you should return an error if the `uninit` function is NULL. Same for other functions below, in case the function is mandatory to implement.

jtguggedal replied:
I can do `if`s instead and return error codes if `NULL` so that it works for `CONFIG_ASSERT=n` as well

jtguggedal commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+
+/**@brief Cloud backend states. */
+enum cloud_state {
+	CLOUD_STATE_DISCONNECTED,
```
I actually thought about dropping it, and let the backends handle their own, internal state themselves.

jtguggedal commented on include/net/cloud.h:
```diff
@@ -0,0 +1,246 @@
+		return -EINVAL;
+	}
+
+	return backend->api->init(backend, handler);
```
Agreed

jtguggedal commented:
@rlubos @metc Thanks for the comments! I'll apply your suggestions, and will ping you when I have updated the API. Hopefully this week.

jtguggedal commented:
I've updated the PR with some changes after your suggestions. The highlights:
- Changed name of the `cloud_resource` enum to `cloud_endpoint`, and added three types of topics that seem to be quite common across cloud solutions: message, state and config. At least Google cloud and AWS have variations of these, with at least one of them representing the digital twin of the device.
- Custom data can now be added to the backend and will be passed as an argument in event callback. The solution follows the pattern of other APIs in Zephyr, with a `cloud_user_data_set( void *user_data)` function. (btw, I think I've made it unnecessarily complicated, so will probably revisit it). 
- Moved from `ASSERT` to `if`s and returning error codes if the backend, API or some of the functions are not implemented.

The cloud codec can be disregarded in this PR, it's mostly a placeholder to ensure functionality, while the changes are handled in #818.

jtguggedal commented:
Added DNM label because of need to rebase the codec on top of #818

rlubos commented:
@jtguggedal I'm slightly confused now, could you clarify which PR depends on which? #818 is built on top of this PR, yet you say that this PR has to be rebased on top of it?

jtguggedal commented:
@rlubos Yeah, I'm confusing myself as well now. The codec in this PR was just intended to be a placeholder for the codec that was underway in #818, and I didn't put much effort into it. Plan was to pull in the other codec when it was ready, but let's just keep it as is in this PR and let #818 improve it.

rlubos commented:
Added @b-gent for docs review.

rlubos commented on applications/asset_tracker/prj.conf:
```diff
@@ -56,5 +59,5 @@ CONFIG_NET_IPV4=n
 CONFIG_NET_TCP=n
-CONFIG_NET_RX_STACK_SIZE=256
-CONFIG_NET_TX_STACK_SIZE=256
+CONFIG_NET_RX_STACK_SIZE=512
```
I'm curious, did anything crash for you so that you needed to increase these stacks?

jtguggedal replied:
Oops, this shouldn't be here. Will remove it.

rlubos commented on applications/asset_tracker/src/cloud_codec/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
```
2019

rlubos commented on applications/asset_tracker/src/main.c:
```diff
@@ -19,6 +18,12 @@
 #endif
 
+#include <net/cloud.h>
+
```
Do we need those empty lines between header inclusions?

rlubos commented on applications/asset_tracker/src/main.c:
```diff
@@ -568,247 +556,91 @@ static void leds_update(struct k_work *work)
 
-	for (u8_t i = 0; i < ARRAY_SIZE(available_sensors); i++) {
-		param.sensor_type = available_sensors[i];
+	k_free(output.buf);
```
I'd consider adding an API to `cloud_codec` for releasing allocated data (something like `cloud_release_data(struct cloud_data *data)`). The app should not need to know what exactly it has to release.

jtguggedal replied:
Good idea, added the suggested API

rlubos commented on applications/asset_tracker/src/main.c:
```diff
@@ -1167,19 +953,68 @@ static void buttons_leds_init(void)
+
+	display_state = LEDS_CONNECTING;
+	ret = cloud_connect(cloud_backend);
+
```
No need for empty line here.

jtguggedal commented on include/net/cloud.h:
```diff
@@ -0,0 +1,308 @@
+ * @brief Cloud event handler function type.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ * @param[in] evt Pointer to cloud event.
```
Fixed, thanks

jtguggedal commented on include/net/cloud.h:
```diff
@@ -0,0 +1,308 @@
+ *	  configuration of the backend to be able to connect its online
+ *	  counterpart.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
```
Fixed them all

jtguggedal commented on subsys/net/lib/cloud/Kconfig:
```diff
@@ -0,0 +1,12 @@
+if CLOUD_API
+
+
+endif # CLOUD
```
Removed the `if`, as it's currently not used for anything

jtguggedal commented:
@rlubos @metc I think I've addressed all of your comments now.
Awaiting doc review by @b-gent

metc commented:
> @rlubos @metc I think I've addressed all of your comments now.

Thank you.

b-gent commented:
I made some edits in the doxygen comments, hopefully CI won't complain. You can review on my commit and then please squash it before merging this.

In general, are there any RST updates planned related to this PR? I'm assuming we should mention the cloud codec somewhere in Asset Tracker documentation and add a new page for Cloud API? If not, then we don't even have a place to display the doxygen-generated doc.

jtguggedal commented:
@b-gent Thanks for the review! Yout changes were good, so I squashed them into the commits where they belong. I added another commit where I attempt to add basic docs for the Cloud API. Unfortunately my documentation skills are limited, so it would be great if you could review. Not sure about the placement of the `cloud.rst` file either.

Regarding the cloud codec, I think we should do the docs in #818 , where the API is subject to change and expansion.

By the way, this API will not be part of the release.

jtguggedal commented:
@hmhalvorsen Re-requested review for the added commit with some initial Cloud API docs

b-gent commented on subsys/net/lib/nrf_cloud/src/nrf_cloud.c:
```diff
@@ -200,3 +201,169 @@ void nrf_cloud_process(void)
+
+static int ping(const struct cloud_backend *const backend)
+{
+	/* TODO: Do only ping, nrf_cloud_process() also checks for input. */
```
IS the TODO intended here?

jtguggedal replied:
Yes, it will be fixed in upcoming PR

b-gent commented on subsys/net/lib/cloud/cloud.rst:
```diff
@@ -0,0 +1,30 @@
+.. _cloud_api:
+
+Cloud API
```
Is this API used in any sample?

jtguggedal replied:
The API is used in the Asset tracker application

b-gent replied:
This information is definitely worth adding in the RST file.

jtguggedal commented:
@b-gent Thanks for reviewing, your commit has been squashed.

jtguggedal commented:
@rlubos I think this one is ready to be merged, although CI seems to be stuck forever on the doc build

rlubos commented:
@jtguggedal Yes, I'm waiting for it, will merge once done.
